File: backend/controllers/ratingController.js
const Rating = require('../models/Rating');
const Recipe = require('../models/Recipe');

const createRating = async (req, res) => {
  try {
    const { rating } = req.body;
    const { recipeId } = req.params;

    const existingRating = await Rating.findOne({
      author: req.user._id,
      recipe: recipeId
    });

    if (existingRating) {
      existingRating.rating = rating;
      await existingRating.save();
    } else {
      await Rating.create({
        rating,
        author: req.user._id,
        recipe: recipeId
      });
    }

    // Update recipe average rating
    const ratings = await Rating.find({ recipe: recipeId });
    const avgRating = ratings.reduce((sum, r) => sum + r.rating, 0) / ratings.length;
    
    await Recipe.findByIdAndUpdate(recipeId, {
      averageRating: avgRating,
      totalRatings: ratings.length
    });

    res.json({ message: 'Rating saved', averageRating: avgRating });
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

const getUserRating = async (req, res) => {
  try {
    const { recipeId } = req.params;
    
    const rating = await Rating.findOne({
      author: req.user._id,
      recipe: recipeId
    });

    res.json({ rating: rating ? rating.rating : null });
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

module.exports = { createRating, getUserRating };

File: springboot-backend/src/main/java/com/foodstagram/controller/RatingController.java
package com.foodstagram.controller;

import com.foodstagram.model.Rating;
import com.foodstagram.model.Recipe;
import com.foodstagram.model.User;
import com.foodstagram.repository.RatingRepository;
import com.foodstagram.repository.RecipeRepository;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/ratings")
public class RatingController {
    private final RatingRepository ratingRepository;
    private final RecipeRepository recipeRepository;

    public RatingController(RatingRepository ratingRepository, RecipeRepository recipeRepository) {
        this.ratingRepository = ratingRepository;
        this.recipeRepository = recipeRepository;
    }

    @PostMapping("/{recipeId}")
    public ResponseEntity<?> createRating(@PathVariable Long recipeId,
                                           @RequestBody Map<String, Integer> body,
                                           @AuthenticationPrincipal User user) {
        Recipe recipe = recipeRepository.findById(recipeId).orElse(null);
        if (recipe == null) return ResponseEntity.notFound().build();

        Optional<Rating> existing = ratingRepository.findByAuthorAndRecipe(user, recipe);
        Rating rating;
        if (existing.isPresent()) {
            rating = existing.get();
            rating.setRating(body.get("rating"));
        } else {
            rating = new Rating();
            rating.setRating(body.get("rating"));
            rating.setAuthor(user);
            rating.setRecipe(recipe);
        }
        ratingRepository.save(rating);

        List<Rating> ratings = ratingRepository.findByRecipe(recipe);
        double avg = ratings.stream().mapToInt(Rating::getRating).average().orElse(0.0);
        recipe.setAverageRating(avg);
        recipe.setTotalRatings(ratings.size());
        recipeRepository.save(recipe);

        return ResponseEntity.ok(Map.of("message", "Rating saved", "averageRating", avg));
    }

    @GetMapping("/{recipeId}")
    public ResponseEntity<?> getUserRating(@PathVariable Long recipeId, @AuthenticationPrincipal User user) {
        Recipe recipe = recipeRepository.findById(recipeId).orElse(null);
        if (recipe == null) return ResponseEntity.notFound().build();
        Optional<Rating> rating = ratingRepository.findByAuthorAndRecipe(user, recipe);
        return ResponseEntity.ok(Map.of("rating", rating.map(Rating::getRating).orElse(null)));
    }
}
